Color mapping tweaks, quantile mode, improved labels, ...

pierstitus opened the pull request:
This patch contains several tweaks to the pseudocolor raster color mapping interface:
- Get rid of leading zeros
- Automatically update labels if the value is manually changed
- Optionally add a unit to the labels
- Make distinction between automatic labels and manual labels, manual labels are skipped by automatic labeling
- Improve sort code, sort automatically after each edit, and remove the sort button 
- Add Quantile based color map classification mode

![color_map_tweaks](https://cloud.githubusercontent.com/assets/4028149/14318609/faa1e3f2-fc0d-11e5-8bfd-2880f31e6f3d.png)

pierstitus committed:
change color map labels to more compact format

pierstitus committed:
automatically change color map labels when editing value

not optimal since autolabel function gets called way too often.

pierstitus committed:
fix too numerous itemChanged events using QTreeWidgetItem subclass

pierstitus committed:
more consistent flags on TreeWidgetItems

pierstitus committed:
use QgsTreeWidgetItem also for sorting color map items

quicker and less code, so enable automatic sorting after every edit

pierstitus committed:
Remove color map sort button, as it is now always automatically sorted

pierstitus committed:
Add option for adding units to color map labels

pierstitus committed:
code style fixes

pierstitus committed:
calculate reasonable number of decimals on classify

pierstitus committed:
Add Quantile based color map creation mode

pierstitus committed:
code style fix (astyle)

gioman commented:
@pierstitus much appreciated, thanks! this is one of those areas of QGIS that really needed some love.

nyalldawson commented:
Sounds great @pierstitus !

First thing we need to do is get these changes passing the CI tests. You'll see https://travis-ci.org/qgis/QGIS/builds/121163401 that your PR has failed the tests on the first build. Click that build and scroll to the bottom, you'll see the note

```
Test results submitted to
  http://tinyurl.com/hhlwvrc
```

in yellow. Open that link (will open the cdash test results dashboard) and you can see there's 2 test failures in red. Clicking the 2 will show that the branch has failed the documentation and sip binding tests.

Click the red 'failed' link on each to see why. In this case you've introduced new public/protected members and classes without doxygen documentation and some have not been added to the corresponding python .sip bindings. You'll need to address that to get the CI tests approving this PR.

When that's done we can move on to the review stage :) Let me know if you need further explanation here!

nirvn commented:
@pierstitus , nice work! Glad someone's refreshing/improving this raster rendering. BTW, if you feel like giving some more love to the pseudocolor renderer, it'd be _great_ to the stretch-related raster toolbar buttons work with it:
![t](https://cloud.githubusercontent.com/assets/1728657/14338177/f3656122-fc9e-11e5-86e5-ffb18e166f75.png)

pierstitus committed:
documentation and sip coverage

pierstitus commented:
Yes, I saw the big scary "All checks have failed" sign.
I'll add some documentation, good convention. Currently there is no documentation in the file at all, but you got to start somewhere. Also for some other new functions it does not complain.

About sip bindings, what is the general idea of what is included in the bindings and what not?

Futhermore I have some difficulties with building: some missing .moc files (e.g. testqgscomposertablev2.moc) and no init_core function in the python _core.so module. It builds well enough to run qgis, but not for unit tests and running the python part. I'm trying again with an updated master, but with the travis ci I would expect that the master branch should compile after every commit, or is that too wishful thinking?

jef-n commented:
@pierstitus are you on ubuntu wily?  It's cmake has a problem (see [#14309](http://hub.qgis.org/issues/14309)).  Downgrading cmake to vivid's version helps.
[CDash - QGIS](http://dash.orfeo-toolbox.org/index.php?project=QGIS) has an overview about to test results of the various builds.  The distribution travis runs on gets a bit more love than the others...

nyalldawson commented:
@pierstitus

The doc tests will fail if any new public or protected members have been introduced without documentation. They were designed to stop the growing number of undocumented classes in the code, so they ignore a hard coded list of members which had no docs at the time the test was introduced. But new stuff will fail without docs, and changes which add docs to existing members are always very welcome!

Re the sip bindings - the rule is if it's public or protected it should be in the bindings. There's a few exceptions though:

- protected variables are ignored, since they can't be added to the bindings
- some members (such as those using templates or complex mapped types) also can't be added. These need to be marked with a doxygen "@note not available in Python bindings" comment so that PyQGIS developers know they aren't available.

Generally it's pretty straightforward to add bindings, the only complications come with transfer of ownership. So we want everything possible added to the bindings.

pierstitus commented:
ok thanks,
I fixed the tests, but now it fails on other tests that have to do with different commits...
So not every master commit seems to pass the tests?

nyalldawson commented on python/gui/raster/qgsrasterminmaxwidget.sip:
```diff
@@ -11,6 +11,9 @@ class QgsRasterMinMaxWidget: QWidget
 
     void setBands( const QList<int> & theBands );
 
+    QgsRectangle extent();
```
The sip bindings also need the doxygen comments - it makes it easier to diff them against the header files.

nyalldawson commented on src/gui/raster/qgsrasterminmaxwidget.h:
```diff
@@ -32,6 +32,11 @@ class GUI_EXPORT QgsRasterMinMaxWidget: public QWidget, private Ui::QgsRasterMin
     void setBands( const QList<int> & theBands ) { mBands = theBands; }
+    /** Return the extent selected by the user.
+        Either an empty extent for 'full' or the current visible extent. */
+    QgsRectangle extent() { QgsRectangle myExtent; return mCurrentExtentRadioButton->isChecked() ? mExtent : myExtent; }
```
Could be simplified to just

````
QgsRectangle extent() { return mCurrentExtentRadioButton->isChecked() ? mExtent : QgsRectangle(); }
````

This would avoid an unnecessary QgsRectangle construction.

nyalldawson commented on src/gui/raster/qgssinglebandpseudocolorrendererwidget.cpp:
```diff
@@ -157,30 +192,58 @@ QgsRasterRenderer* QgsSingleBandPseudoColorRendererWidget::renderer()
-    delete currentItem;
+    myCurrentItem = mColormapTreeWidget->topLevelItem( i );
+    //If the item is null or does not have a pixel values set, skip
+    if ( !myCurrentItem || myCurrentItem->text( 0 ) == "" )
```
Instead of `myCurrentItem->text( 0 ) == ""` use `myCurrentItem->text( 0 ).isEmpty()` (much more efficient)

nyalldawson commented on src/gui/raster/qgssinglebandpseudocolorrendererwidget.cpp:
```diff
@@ -157,30 +192,58 @@ QgsRasterRenderer* QgsSingleBandPseudoColorRendererWidget::renderer()
+      label = myCurrentItem->text( 0 ) + unit;
+    }
+
+    if ( myCurrentItem->text( 2 ) == "" || myCurrentItem->text( 2 ) == label || myCurrentItem->foreground( 2 ).color() == QColor( Qt::gray ) )
```
same here:" `myCurrentItem->text( 2 ).isEmpty()`

nyalldawson commented on src/gui/raster/qgssinglebandpseudocolorrendererwidget.cpp:
```diff
@@ -157,30 +192,58 @@ QgsRasterRenderer* QgsSingleBandPseudoColorRendererWidget::renderer()
-  int myCurrentIndex = 0;
+  bool discrete = mColorInterpolationComboBox->currentText() == tr( "Discrete" );
+  QStringList allSuffixes;
+  QString label = "";
```
Just go `QString label` - no need to initialize to an empty string, that's already done by QString constructor.

nyalldawson commented on src/gui/raster/qgssinglebandpseudocolorrendererwidget.cpp:
```diff
@@ -190,41 +253,69 @@ void QgsSingleBandPseudoColorRendererWidget::on_mSortButton_clicked()
+  QgsTreeWidgetItem* newItem = new QgsTreeWidgetItem( mColormapTreeWidget );
+  newItem->setText( 0, "0" );
+  newItem->setBackground( 1, QBrush( QColor( Qt::magenta ) ) );
+  newItem->setText( 2, "" );
```
`newItem->setText( 2, QString() );`

nyalldawson commented on src/gui/raster/qgssinglebandpseudocolorrendererwidget.cpp:
```diff
@@ -260,6 +351,38 @@ void QgsSingleBandPseudoColorRendererWidget::on_mClassifyButton_clicked()
+    int bandNr = mBandComboBox->itemData( bandComboIndex ).toInt();
+    //QgsRasterHistogram myRasterHistogram = mRasterLayer->dataProvider()->histogram( bandNr );
+
+    double myMin = std::numeric_limits<double>::quiet_NaN();
```
avoid use of `my` prefixes for variables. They have been confusingly used in very old QGIS code as both local and private member variables. The convention is `mVariable` for member variables, and just `variable` for local variables.

nyalldawson commented on src/gui/raster/qgssinglebandpseudocolorrendererwidget.cpp:
```diff
@@ -335,19 +458,26 @@ void QgsSingleBandPseudoColorRendererWidget::on_mClassifyButton_clicked()
   QVector<QColor>::const_iterator color_it = entryColors.begin();
 
+  // calculate a reasonable number of decimals to display
+  double maxabs = log10( std::max<double>( std::fabs( max ), std::fabs( min ) ) );
```
This will need to use `qAbs` instead of `std::fabs` - otherwise it breaks the build on some platforms

nyalldawson commented on src/gui/raster/qgssinglebandpseudocolorrendererwidget.cpp:
```diff
@@ -335,19 +458,26 @@ void QgsSingleBandPseudoColorRendererWidget::on_mClassifyButton_clicked()
 
+  // calculate a reasonable number of decimals to display
+  double maxabs = log10( std::max<double>( std::fabs( max ), std::fabs( min ) ) );
+  int nDecimals = round( std::max<double>( 3.0 + maxabs - log10( max - min ), maxabs <= 6.0 ? maxabs + 0.49 : 0.0 ) );
```
This will also need to use `qRound`, `qMax` in place of the std round/max

nyalldawson commented on src/gui/raster/qgssinglebandpseudocolorrendererwidget.cpp:
```diff
@@ -335,19 +458,26 @@ void QgsSingleBandPseudoColorRendererWidget::on_mClassifyButton_clicked()
+    newItem->setText( 0, QString::number( *value_it, 'g', nDecimals ) );
     newItem->setBackground( 1, QBrush( *color_it ) );
-    newItem->setText( 2, QString::number( *value_it, 'f' ) );
+    newItem->setText( 2, "" );
```
`ewItem->setText( 2, QString() );`

nyalldawson commented on src/gui/raster/qgssinglebandpseudocolorrendererwidget.cpp:
```diff
@@ -576,10 +710,32 @@ void QgsSingleBandPseudoColorRendererWidget::on_mColormapTreeWidget_itemDoubleCl
   }
   else
   {
+    if ( column == 2 )
```
It'd be a good idea to move these hardcoded column numbers to a private enum and use the enum values instead of the raw integers, just in case the columns are shifted around in future

nyalldawson commented:
> I fixed the tests, but now it fails on other tests that have to do with different commits...
So not every master commit seems to pass the tests?

Not sure what happened there - I reran the failing tests and they passed OK.

nyalldawson commented:
@pierstitus Here's my thoughts after testing this:

In general, it looks and works great. These are fantastic improvements! I'd suggest moving the label suffix widget up to sit below the color interpolation combo box and adding a label to its left. At the moment it's not immediately obvious what this empty text box does.

There's something up with how the "quantile" method calculates breaks. If I set it to 2 classes and discrete I'd expect half the pixels to be shaded one color and the other half the other color. But instead I see only a handful of pixels shaded in the first color and the majority in the second class color. I suspect this is due to how the renderer treats the first break as a " <= " value.

I also think the label for the first class when in discrete mode should be " <= ", not < . But it's no immediately obvious from the raster shader code if this is really the case.

pierstitus committed:
implement comments from nyalldawson

pierstitus committed:
Merge branch 'master' into pseudocolor_tweaks

pierstitus commented:
@nyalldawson Thanks for the comments, many Qt beginners mistakes :)

As for the discrete mode, that's actually broken with all classification modes, as none of them takes into account the render mode. Also discrete mode is a bit peculiar in that the first class <= the first value but the last class only up to the last value. Values above the last colormap value are not rendered. This behaviour is also different than the discrete mode in the color gradient editor, which duplicates the last color, while to be consistent with the renderer the first color should be duplicated. (at least in the current version, I don't know if your redesigned editor does it different?)

There would be two ways to make it symmetric:
 - first and last class have <= and > respectively, this means the min and max are lost. Can be obtained by setting the last value to 'inf'
 - add an extra class at the start, transparent color, for defining the minimum. This is more similar to the gradient editor.
I'd think the first option is most common, and easiest to implement.

It is a bit of a mess to handle all the discrete gradients, also because colorbrewer doesn't add the extra class. Maybe it would be good to move some functionality to the QgsVectorColorRampV2 classes.

About the empty box, I agree it's not immediately clear, but I was a bit reluctant to take yet an extra line of screen real estate.

pierstitus committed:
Different color map classification handling in discrete mode

pierstitus committed:
Color map label on different line

pierstitus committed:
fix colors for continuous classification with non equally spaced stops

pierstitus committed:
fix handling discrete colormaps in continuous mode

pierstitus commented:
Ok, I implemented the discrete classify cases, as in the first option. 

Also I now see that continuous mode it didn't take the right colors when the gradient itself is not equally distributed. That's fixed now.

Another thing that is not very clear is that the accuracy and extent options influence the Quantile classification, any ideas?

pierstitus committed:
limit number of classes to minimum 2 and remove special cases for < 2

pierstitus committed:
forgot unit

pierstitus committed:
forgot cast to double on divide

pierstitus commented:
Again it failed passing tests because the master branch failed. I triggered a new build with another push, should be ok now.

@nirvn interesting idea, but personally I think it makes more sense to have the styling in a dock, as @NathanW2 is implementing https://github.com/qgis/QGIS/pull/2967. If you have that the limited control of these buttons doesn't add so much I'd say.

nirvn commented:
@pierstitus , the nice thing with the toolbar buttons (which wouldn't be replaced by the styling dock) is that, within a single toolbar click, you can arrange the color stretch to your canvas extend. That's very useful when working with LiDAR data, etc. The button saves you from having to:
- open a style panel (whether as a dock or in a window),
- click on [ load ] min/max values button
- click on classify

It combines all of that into a single toolbar click. Try it with a singleband gray raster, it's magical :smile:.

pierstitus committed:
[raster] Fix pseudocolor clip option (setting was lost in the pipeline)

Also disable checkbox for exact and discrete mode, as it isn't used there.

pierstitus commented:
Added other related bugfix: the "Clip" option was not working at all. Also the clip option only has effect on linear interpolation so the checkbox is now disabled when exact or discrete are selected.

Also I found that the implementation of the discrete mode is different than it's documentation suggests. That says "Assigns the color of the lower class for every pixel between two class breaks." (in qgscolorrampshader.h), which is opposite to what happens now. See also http://hub.qgis.org/issues/13995

Besides this I rewrote the QgsColorRampShader class to drastically improve the render speed, should I add that commit here or do a separate pull request?

NathanW2 commented:
@pierstitus nice work!  

I really like removing the UI stuff that isn't needed anymore

m-kuhn commented:
The rules for documentation are: any new method must be documented. There are still a lot of leftovers from before this rule was introduced for which docs are written incrementally on a volunteer base.

The rules for sip (python) bindings are: any new method must come with python bindings unless there is a *good reason ™* not to write one. There is `./scripts/sipdiff [header file here]` which helps with writing the sip files (if you are used to vim).

I think the moc problem is related to the cmake version in ubuntu wily and should be fixed meanwhile https://bugs.launchpad.net/ubuntu/+source/cmake/+bug/1527722. If you still get this error after updating cmake from the ubuntu repository, please leave a comment here https://github.com/qgis/QGIS/pull/2679.

We try to keep master green but we sometimes fail due to external changes or because of failures from our side.

Sorry for the delay, I didn't notice that there are unanswered questions here.

pierstitus committed:
Reduce color map display decimals to max 15

Using 16 decimals gives rounding issues, (e.g. 64.4 displayed as 64.40000000000001)

pierstitus commented:
@m-kuhn thanks for clarifying. The moc problem I solved for now by manually generating them with a bash loop, since Ubuntu 16.04 is nearly there anyway.

Again the checks failed on a completely different part, only on osx. Anyway the other test succeed :)

@nyalldawson Is there still anything that requires changes?

m-kuhn commented:
No worries, this test can be safely ignored. I'll probably turn it off since at the moment it's more distracting than any useful.

pierstitus commented:
Can someone please merge this, or is there something that needs attention?

pierstitus commented:
@NathanW2, did you find some time to test this patch? It would be nice to give merge it to get some broader testing.
Thanks!

nyalldawson commented:
Working on this now

nyalldawson commented:
I've manually merged all this. Thank you for your hard work - this is great stuff!

pierstitus commented:
Cool, always nice to be able to give something back!

A quick note about Quantile Mode that is not obvious, for documentation or improvement later:
Quantile Mode uses the Extent and Accuracy settings, and not the Min and Max fields.

Anyway it could be nice to replace the QgsRasterMinMaxWidget with something custom so the interface could be improved.
